cmd/kubectl-gadget/utils/trace.go

```diff
@@ -463,95 +466,157 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
+			// 3. Or time out is fired.
+			// NOTE In case 2 and 3, it exists, at least, one trace we did not deal
+			// with.
+			switch event.Type {
+			case watch.Deleted:
+				// If for some strange reasons (e.g. users deleted a trace during this
+				// operation) a trace is deleted, we need to take care of this by
+				// decrementing the tracesToWatch.
+				// Otherwise we would still wait for the old number and we would
+				// timeout.
+				tracesToWatch--
+
+				trace, ok := event.Object.(*gadgetv1alpha1.Trace)
+				if !ok {
+					return false, fmt.Errorf("Received object cannot be cast to Trace: %v", event.Object)
+				}
 
-			nodeErrors[trace.Spec.Node] = trace.Status.OperationError
+				traceName := trace.ObjectMeta.Name
+
+				// We also remove it from the maps to avoid returning a deleted trace
+				// and timeing out.
+				delete(satisfiedTraces, traceName)
+				delete(erroredTraces, traceName)
+
+				return false, nil
+			case watch.Modified:
+				// We will deal with this type of event below
+			case watch.Error:
+				// Deal particularly with error.
+				return false, fmt.Errorf("Received event is an error one: %v", event)
+			case watch.Added:
+				// createTraces() creates traces synchronously.
+				// So, if a watch.Added event occurs it means there is a problem (e.g.
+				// the user creates a trace by snooping on the traceID of existing
+				// traces).
+				return false, fmt.Errorf("No traces with the given traceID (%s) should be created", traceID)
+			default:
+				// We are not interested in other event types.
+				return false, nil
+			}
 
-			return false, nil
-		}
+			trace, ok := event.Object.(*gadgetv1alpha1.Trace)
+			if !ok {
+				return false, fmt.Errorf("Received object cannot be cast to Trace: %v", event.Object)
+			}
 
-		// If the trace does not satisfy the condition function, we are not
-		// interested.
-		if !conditionFunction(trace) {
-			return false, nil
-		}
+			if trace.Status.OperationWarning != "" {
+				// The trace can have a warning but satisfies conditionFunction.
+				// So, we do not add it to the map here.
+				nodeWarnings[trace.Spec.Node] = trace.Status.OperationWarning
+			}
 
-		watchedTracesNumber++
-		// If the current trace matches our filter we add it to the list of trace
-		// we will return.
-		returnedTraces.Items = append(returnedTraces.Items, *trace)
+			if trace.Status.OperationError != "" {
+				erroredTraces[trace.ObjectMeta.Name] = trace
 
-		return watchedTracesNumber == tracesNumber, nil
-	})
-	cancel()
+				// If the trace satisfied the function, we do not care now because it
+				// has an error.
+				delete(satisfiedTraces, trace.ObjectMeta.Name)
+
+				return false, nil
+			}
+
+			// If the trace does not satisfy the condition function, we are not
+			// interested.
+			if !conditionFunction(trace) {
+				return false, nil
+			}
+
+			satisfiedTraces[trace.ObjectMeta.Name] = trace
+
+			return len(satisfiedTraces)+len(erroredTraces) == tracesToWatch, nil
```

Comment:
Shouldn't it be `len(satisfiedTraces)+len(erroredTraces) == tracesNumber` ? (`satisfiedTraces` and `erroredTraces` can be filled with traces before the watch operation).